### ISSUE TITLE
Get source code location in trace `call` messages

### ISSUE THREAD

**Issue by sidkshatriya** [erlang/otp#5297]
trace `call` messages have the following format

`{trace, Pid, call, {M, F, Args}}`

See https://erlang.org/doc/man/erlang.html#trace-3 

This allows you to get the parameters of the function call easily. By using `return_trace` in a call match specification you can also get the return value of the function. These are very useful. But sometimes a function may have many invocations of a particular function (with the same arity) and it would be useful to disambiguate them.

__One feature that would be very useful is where exactly is the function call coming from in source code?__

So something like this would be useful 

`{trace, Pid, call_with_loc, {M, F, Args}, CodeLocation}`

Where CodeLocation could be something like `cowboy_http.erl:45` or `phoenix.ex:900`.

Questions:
- Can we accomplish something similar like this today?
- If not, would this be a useful feature to implement? Does it require a lot of changes to the Erlang/OTP code?
- Would the core team be interested in implementing something like this?
- How could one go about implementing such a feature?

**Comment by jhogberg** [comment]
> I'll obviously need to get into the C-code to understand how things work. I'm not familiar with beam internals. Are there some good resources I can read to get a grip on the fundamentals?

The documentation we have mostly concerns how breakpoints are set (e.g. [Tracing.md](https://github.com/erlang/otp/blob/master/erts/emulator/internal_doc/Tracing.md) or the [tracing/nif loading section in BeamAsm.md](https://github.com/erlang/otp/blob/master/erts/emulator/internal_doc/BeamAsm.md#tracing-and-nif-loading)) since that's the trickiest part, but that's not very relevant here. I'll make an effort to expand the documentation.

In the meantime, I'll be happy to answer your questions. :)

> Could the following work (pseudo-code) ? ... snip ...

It would work but it would have a slight overhead on _all_ calls since it's impossible to know whether the callee is traced. The main benefit of our current method is that is has no overhead when tracing is inactive, and that's something we don't want to lose.

**Comment by sidkshatriya** [comment]
Thanks. 

> It would work but it would have a slight overhead on all calls since it's impossible to know whether the callee is traced.

That makes sense. What if we stored _just_ the address (of the bytecode) of the last call site somewhere in the process data structure. Later, if we need to calculate the source line we can use this information. This would add minimal overhead and could only happen if we switched "caller_line" tracing for that process. One could have this switch on a module level also if that makes more sense. Alternatively this could be a compiler switch and would insert a code snippet before every call site but that is not a very appealing option because you would need to compile your code in advance with that switch. However, its better than nothing...

**Why is this important?**
For my use case having accurate call site information under all scenarios is extremely important. Using the debugger is extremely slow and impractical. Processes often start timing out when things are stopped in the interactive debugger. 

Tracing is great but one of its limitations is that you don't know where the function call is coming from! In Elixir there is so much meta-programming and functions can be large. Just knowing the name of caller is not enough. It would be great to know the exact source location. Then tracing essentially becomes almost as good as debugging because (a) functions are everywhere and you just follow your eyes to the next function in the code base to see what value it returned (you'll now have the exact location of the next function call). (b) Since there is no mutation in erlang/elixir code anyways you don't really need to know what happened between function calls.

TL; DR - Source code locations while tracing function calls will give you a super fast debugger. When you're doing tracing with `return_trace` the system could do bad things anyways.

**Comment by jhogberg** [comment]
> While trying to implement this feature will I need to worry about the JIT? In other words will the changes to the interpreter need to be to re-implemented somehow in the JIT or will the changes carry through?

The `caller_line` option will work independently of everything else. Essentially what you need to do is change [the call to `erts_find_function_from_pc`](https://github.com/erlang/otp/blob/e71c38afcd2e87826ee1c3f9dd1edeae0c1daa0a/erts/emulator/beam/erl_db_util.c#L2741) to `erts_lookup_function_info`, and then send the line information over.

It won't be as accurate without the other changes but it should work just fine on its own.

> The emulator flag approach to get totally accurate caller information (in the face of tail recursion) I'm sure adds more complexity to this feature. What might be a sketch of an approach that could be taken here?

Yep, it's quite involved and I think we'll want to do this part ourselves. For what it's worth:

I'd start out by extending the `return_trace`/`exception_trace` frames so that they contain the current caller address (filling in the current return address when the feature is disabled).

In the interpreter we'll need to use different instructions based on whether this feature is enabled. I'd add a loader predicate that returns whether the feature is enabled, and conditionally rewrite all call instructions to ones that set the caller address when it holds true. The relevant documentation can be found in [beam_makeops.md](https://github.com/erlang/otp/blob/master/erts/emulator/internal_doc/beam_makeops.md#defining-transformation-rules).

In the JIT I'd change the calling convention to keep the caller address in a predefined machine register and only set it in the process structure when hitting a breakpoint, scheduling out, or when returning into a `return_trace` frame. It would need to be loaded in a few edge cases too (e.g. BIF trapping), but for the most part we wouldn't need to care about it since it only needs to be valid in function prologues.

That's the gist of it anyway. I really like the possibilities this opens up (`rr` clone for Erlang? :smiley:) so I'll see if we can get this prioritized for OTP 25.
